src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +187,1485 @@
+
     """
-    for i in range(0, transitions):
-        if timestep < (snapshot_years[i+1] - snapshot_years[0]):
-            return i
+    suffix = utils.make_suffix_string(args, 'results_suffix')
+    output_dir = os.path.join(args['workspace_dir'], 'output')
+    intermediate_dir = os.path.join(args['workspace_dir'], 'intermediate')
+    taskgraph_cache_dir = os.path.join(intermediate_dir, 'task_cache')
 
+    utils.make_directories([output_dir, intermediate_dir, taskgraph_cache_dir])
 
-def s_to_timestep(snapshot_years, snapshot_idx):
-    """Convert snapshot index position to timestep.
+    try:
+        n_workers = int(args['n_workers'])
+    except (KeyError, ValueError, TypeError):
+        # KeyError when n_workers is not present in args
+        # ValueError when n_workers is an empty string.
+        # TypeError when n_workers is None.
+        n_workers = -1  # Synchronous mode.
+    task_graph = taskgraph.TaskGraph(
+        taskgraph_cache_dir, n_workers, reporting_interval=5.0)
+
+    snapshots = _extract_snapshots_from_table(args['landcover_snapshot_csv'])
+
+    # Phase 1: alignment and preparation of inputs
+    baseline_lulc_year = min(snapshots.keys())
+    baseline_lulc_path = snapshots[baseline_lulc_year]
+    baseline_lulc_info = pygeoprocessing.get_raster_info(
+        baseline_lulc_path)
+    min_pixel_size = numpy.min(numpy.abs(baseline_lulc_info['pixel_size']))
+    target_pixel_size = (min_pixel_size, -min_pixel_size)
+
+    transition_years = set()
+    analysis_year = int(args['analysis_year'])
+
+    base_paths = [baseline_lulc_path]
+    aligned_lulc_paths = {}
+    aligned_paths = [os.path.join(
+        intermediate_dir,
+        f'aligned_lulc_baseline_{baseline_lulc_year}{suffix}.tif')]
+    aligned_lulc_paths[baseline_lulc_year] = aligned_paths[0]
+    for snapshot_year in snapshots:
+        if snapshot_year == baseline_lulc_year:
+            continue
+        base_paths.append(snapshots[snapshot_year])
+        transition_years.add(snapshot_year)
+        aligned_paths.append(
+            os.path.join(
+                intermediate_dir,
+                f'aligned_lulc_snapshot_{snapshot_year}{suffix}.tif'))
+        aligned_lulc_paths[snapshot_year] = aligned_paths[-1]
+
+    prices = None
+    if ('do_economic_analysis' in args and
+            args['do_economic_analysis'] not in ('', None)):
+        if args.get('use_price_table', False):
+            prices = {
+                year: values['price'] for (year, values) in
+                utils.build_lookup_from_csv(
+                    args['price_table_path'], 'year').items()}
+        else:
+            inflation_rate = float(args['inflation_rate']) * 0.01
+            annual_price = float(args['price'])
 
-    Args:
-        snapshot_years (list): list of snapshot years.
-        snapshot_idx (int): index of snapshot
+            try:
+                max_year = max(transition_years.union(set([analysis_year])))
+            except ValueError:
```

Comment:
Curious why the try/except here rather than `if transition_years... else...`? I've heard that using try/except is more pythonic. Is that true and why would you choose one over the other for a case like this?

Reply:
It's true!  It is accepted and even recommended to use reasonable and expected exceptions for control flow, which generally follows the form of:

```
try:
    the thing you would like to have work as expected
except possible_error:
    the backup action instead if it doesn't work
```

So in the case of the above, the assumption is `len(transition_years) > 0`.  In this case, that is an easy enough thing to specify, and so in this case I actually think that your suggested approach is far more direct.  The `ValueError` I had here does not make it immediately clear what the exception was and why it happened, whereas `if transition_years` is much clearer.

Exceptions are especially useful because they allow us to assume that the thing we want to work will work as expected, and yet we are able to express backup or cleanup solutions as needed in a fairly compact syntax.  Exceptions can also be useful for breaking out of multiple nested loops or function calls, ensuring something happens regardless of success or error state, or just jumping into a possibly-expensive operation without worrying ahead of time whether it might fail.

Reply:
Cool, thanks for the explanation!